### ISSUE TITLE
cupy copyto disallows numpy arrays

### ISSUE THREAD

**Issue by hmaarrfk** [cupy/cupy#8165]
### Description

I understand that memory management in frowned up in Python, but sometimes you just have to do it.

`cp.asarray` seems to use a memory pool, but a mechanism to specify the data buffer where data should go seems like it would be beneficial, even in limited usecases.

could `cp.copyto` be expanded to provide at least preliminary support for well behaved numpy arrays (aligned, contiguous, anything else you can dream of)

### Additional Information

sometimes in troubleshooting you just want to know that you aren't hitting up any memory management issues.

I've been playing around with lazyarrays, and the ability to dispatch routines like `copyto` is pretty helpful.

**Comment by leofang** [comment]
Q: Would it be possible for you to just use `cupy.asnumpy()`? Do you really need the matching signature? If so, how about this? 
```python
def copyto(dst, src):
    if isinstance(dst, numpy.ndarray):
        cupy.asnumpy(src, out=dst)
    elif isinstance(dst, cupy.ndarray):
        cupy.copyto(dst, src)
    else:
        raise NotImplementedError
```

**Comment by hmaarrfk** [comment]
```
a = np.zeros(3)
buff = cp.zeros(3)
c = cp.asarray(a, out=buff)
```

i guess i'm looking for `numpy -> cupy` transfer., like the example above. Matching syntax doesn't matter for me (so much)

**Comment by leofang** [comment]
Ah OK. I think what you need is already there, it's just that it's currently guarded behind an env var:
```python
# run this script with the env var CUPY_EXPERIMENTAL_SLICE_COPY=1
import cupy as cp
import numpy as np

a = cp.empty((4, 5))
b = np.arange(20, dtype=a.dtype).reshape(4, 5)
a[:] = b  # copy numpy -> cupy
print(a)
```
https://docs.cupy.dev/en/stable/reference/environment.html#envvar-CUPY_EXPERIMENTAL_SLICE_COPY

I guess what happened was that after we added the initial support we never bothered to revisit this nice feature. Let us know if this works in your use cases and if this is enough for you.

**Comment by hmaarrfk** [comment]
Yes this would do it for me.

I would also say that:

```
a[...] = b
```
doesn't work, even if `CUPY_EXPERIMENTAL_SLICE_COPY=1`

I often use the `...` syntax.

**Comment by leofang** [comment]
> I often use the `...` syntax.

I agree, @asi1024 do you have any thoughts on how we can improve this?

**Comment by hmaarrfk** [comment]
I somewhat want to check on the logic that is used. I spent a bit of time looking through the cython code.

The codepath for `cupy.asarray` does the following:

1. Explains that there are bugs in not using pinned memory
2. Creates a pinned buffer
3. Copies the user's data to the pinned buffer
4. calls `copy_from_host_async` 

The codepath that gets triggered by experimental copy skips creating a pinned buffer. 
https://github.com/cupy/cupy/blob/f64337981cb445ea3f19d4aee19cd4484f30faab/cupy/_core/core.pyx#L1582

Is that going to be a problem?

**Comment by leofang** [comment]
Using a staging buffer was meant to avoid potential race condition (say you modify the NumPy array while copying to GPU) by sacrificing a bit of perf (now we copy twice, from np to staging to cp), but I feel if someone is courageous enough to use an env var, they know what they're doing and we should not need to overcomplicate things. Also, CUDA driver also has an internal staging if the source array is not pinned (i.e. normal NumPy array), so we should just let the driver handle it.

**Comment by hmaarrfk** [comment]
thanks. I think I need to review the meaning of the words like async by reading through some of implications they have.

I guess this issue can be closed. I somewhat understand why you don't allow this API.

**Comment by leofang** [comment]
@pentschev I recall you added the support for `CUPY_EXPERIMENTAL_SLICE_COPY` in #2079. Do you recall why we didn't also allow using ellipsis here?

**Comment by pentschev** [comment]
> @pentschev I recall you added the support for `CUPY_EXPERIMENTAL_SLICE_COPY` in #2079. Do you recall why we didn't also allow using ellipsis here?

I think we never discussed using ellipsis there because we wanted _one_ way of permitting that copy that _is not_ a regular coding pattern, so to speak. Personally, I don't see ellipsis copies that often in production code so I think it would be fine to also add that support if there's real need/desire for it, but it would probably be good to have @kmaehashi weigh in on this matter.
